### PR TITLE
gh-102251: Disable non-rerunnable test in test_import

### PR DESCRIPTION
**Pull request by erlend-aasland**
Alternative to gh-104796; proposed as this PR has a much smaller diff and is less intrusive.

* Issue: gh-102251

### CODE REVIEW

**Comment by sunmy2019** [comment]
I like this solution more. 👍

**Comment by erlend-aasland** [comment]
> I like this solution more. 👍

Yes; my rationale is:

Keep the diff down; keep the solution as simple as possible. We want to be able to easily remove the `no_rerun()` hack when the original problem has been addressed. With this solution we can simply remove the decorator from the affected test function. Also, it limits the decorator to this file only, and adds a relatively big warning for the programmer.

**Review by Yhg1s — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Comment on ref leak run:

For all buildbots, only `test_peg_generator` is leaking, except:

- `buildbot/s390x RHEL8 Refleaks PR`: `test_import` leaks [1, 1, 1] memory blocks! [Link](https://buildbot.python.org/all/#/builders/558/builds/1328/steps/5/logs/warnings__61_)

Suggesting to land this PR as a start, then focus on the RHEL8/s390x issue.

File: Lib/test/test_import/__init__.py [hunk 1]
```diff
@@ -2434,6 +2449,7 @@ def test_with_reinit_reloaded(self):
     # Also, we test with a single-phase module that has global state,
     # which is shared by all interpreters.
 
+    @no_rerun(reason="rerun not possible; module state is never cleared (see gh-102251)")
```
Comment (sunmy2019) [inline Lib/test/test_import/__init__.py]: This test is not leaking.
```suggestion
```

**Review by sunmy2019 — COMMENTED** [review]
Other parts LGTM

File: Lib/test/test_import/__init__.py (hunk 1 above)
Comment (erlend-aasland) [reply Lib/test/test_import/__init__.py]: It is; please see the previous buildbot run.

Comment (erlend-aasland) [reply Lib/test/test_import/__init__.py]: See also ref leak run on Ubuntu container:

<details>

```console
root@9ad8da62d4a6:/src# git checkout e5baa024265d9489da7e354e25021f82c34ade3b
Note: switching to 'e5baa024265d9489da7e354e25021f82c34ade3b'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at e5baa02426 Doh
root@9ad8da62d4a6:/src# git show HEAD
commit e5baa024265d9489da7e354e25021f82c34ade3b (HEAD)
Author: Erlend E. Aasland <erlend.aasland@protonmail.com>
Date:   Fri Jun 23 13:04:24 2023 +0200

    Doh

diff --git a/Lib/test/test_import/__init__.py b/Lib/test/test_import/__init__.py
index 938e7e636a..1114ae2218 100644
--- a/Lib/test/test_import/__init__.py
+++ b/Lib/test/test_import/__init__.py
@@ -2449,7 +2449,7 @@ def test_with_reinit_reloaded(self):
     # Also, we test with a single-phase module that has global state,
     # which is shared by all interpreters.
 
-    #@no_rerun(reason="rerun not possible; module state is never cleared (see gh-102251)")
+    @no_rerun(reason="rerun not possible; module state is never cleared (see gh-102251)")
     @requires_subinterpreters
     def test_basic_multiple_interpreters_main_no_reset(self):
         # without resetting; already loaded in main interpreter
root@9ad8da62d4a6:/src# pushd
/build /src
root@9ad8da62d4a6:/build# ./python -m test -R : test_import
0:00:00 load avg: 0.37 Run tests sequentially
0:00:00 load avg: 0.37 [1/1] test_import
beginning 9 repetitions
123456789
.........
test_import leaked [10, 10, 10, 10] references, sum=40
test_import leaked [8, 8, 8, 8] memory blocks, sum=32
test_import failed (reference leak)

== Tests result: FAILURE ==

1 test failed:
    test_import

Total duration: 29.1 sec
Tests result: FAILURE
root@9ad8da62d4a6:/build# pushd
/src /build
root@9ad8da62d4a6:/src# git checkout fix-test-import-leaks-alt
Previous HEAD position was e5baa02426 Doh
Switched to branch 'fix-test-import-leaks-alt'
Your branch is up to date with 'origin/fix-test-import-leaks-alt'.
root@9ad8da62d4a6:/src# git show HEAD
commit 0784bc19b8ba5b0f07fd6c15f35e77e10fd4d20f (HEAD -> fix-test-import-leaks-alt, origin/fix-test-import-leaks-alt)
Author: Erlend E. Aasland <erlend.aasland@protonmail.com>
Date:   Fri Jun 23 14:17:10 2023 +0200

    Also skip reruns for test_basic_multiple_interpreters_deleted_no_reset()

diff --git a/Lib/test/test_import/__init__.py b/Lib/test/test_import/__init__.py
index 1114ae2218..0a53b4b182 100644
--- a/Lib/test/test_import/__init__.py
+++ b/Lib/test/test_import/__init__.py
@@ -2518,6 +2518,7 @@ def test_basic_multiple_interpreters_main_no_reset(self):
         #  * m_copy was copied from interp2 (was from interp1)
         #  * module's global state was updated, not reset
 
+    @no_rerun(reason="rerun not possible; module state is never cleared (see gh-102251)")
     @requires_subinterpreters
     def test_basic_multiple_interpreters_deleted_no_reset(self):
         # without resetting; already loaded in a deleted interpreter
root@9ad8da62d4a6:/src# pushd
/build /src
root@9ad8da62d4a6:/build# ./python -m test -R : test_import
0:00:00 load avg: 0.32 Run tests sequentially
0:00:00 load avg: 0.32 [1/1] test_import
beginning 9 repetitions
123456789
.........

== Tests result: SUCCESS ==

1 test OK.

Total duration: 27.5 sec
Tests result: SUCCESS
root@9ad8da62d4a6:/build# 
```

</details>

Comment (sunmy2019) [reply Lib/test/test_import/__init__.py]: Isn't the Ubuntu ref leak run only indicating `test_basic_multiple_interpreters_deleted_no_reset` leaking? (edited)

Comment (sunmy2019) [reply Lib/test/test_import/__init__.py]: As for `test_basic_multiple_interpreters_main_no_reset`,
skipping it resulting the same leakage with `main` 6a80664ef


```
./python -m test test_import -R: -j 1
0:00:00 load avg: 0.12 Run tests in parallel using 1 child processes
0:00:15 load avg: 0.31 [1/1/1] test_import failed (reference leak)
beginning 9 repetitions
123456789
.........
test_import leaked [10, 10, 10, 10] references, sum=40
test_import leaked [8, 8, 8, 8] memory blocks, sum=32

== Tests result: FAILURE ==

1 test failed:
    test_import

Total duration: 15.4 sec
Tests result: FAILURE

Comment (erlend-aasland) [reply Lib/test/test_import/__init__.py]: Ah, yeah, of course; thanks for the heads up. I'll remove the decorator from `test_basic_multiple_interpreters_main_no_reset` before landing. Tested out locally; all ok.

**Review by brettcannon — APPROVED** [review]

**Comment by vstinner** [comment]
> @no_rerun(reason="rerun not possible; module state is never cleared (see gh-102251)")

IMO @no_rerun() is bad pattern and should be avoided or removed. They are many tests with side effects which cannot be cancelled because there is no API for that. For example, in the past, there were tests registering codecs and error handlers.

The usual fix is to **run such test in subprocesses**.

For example, test_audit runs subprocesses on functions registering audit hooks, since there is no function to unregister a audit hook **by design**.

**Comment by erlend-aasland** [comment]
@vstinner: I agree that it is a bad pattern. This is a temporary fix only. See my comment earlier in this PR:

> We want to be able to easily remove the no_rerun() hack when the original problem has been addressed.

**Comment by erlend-aasland** [comment]
> The usual fix is to run such test in subprocesses.

ISTM we should rework parts of `test_import` to follow this pattern.

**Comment by Yhg1s** [comment]
This needs a backport to 3.12 to resolve the test_import leak that only shows up when you run `python -m test -R:: test_import` instead of `python -m test -R :: test_import` (note the space between `-R ::`). Some of the buildbots run with `-R::`.

**Comment by vstinner** [comment]
If a test makes a change which cannot be undone, one option is to run it in a subprocess.

Another option is to add an API to undo the change. For example, codecs.unregister() was added to fix tests adding codecs.

See issue #108963 which is a similar issue.

**Comment by vstinner** [comment]
Anyway, thanks for removing this ugly workaround :-(

```
    def setUpClass(cls):
        if '-R' in sys.argv or '--huntrleaks' in sys.argv:
            # https://github.com/python/cpython/issues/102251
            raise unittest.SkipTest('unresolved refleaks (see gh-102251)')
```

**Comment by erlend-aasland** [comment]
Oh, I totally forgot about this. I still think Victor's proposal[^1] should be applied; this PR is a _workaround_, not a fix :)


[^1]: _"The usual fix is to run such test in subprocesses."_ — V. Stinner.

**Pull request merged**